From 30ec4912d5c51510571d93e899d3a424a32c4950 Mon Sep 17 00:00:00 2001
From: aken_liu <aken_liu@accton.com.tw>
Date: Wed, 21 Nov 2018 13:12:52 +0800
Subject: [PATCH] workaround for hard-reset in R0A blade check the blade
 version and then execute the reset function.  Hard reset: R0A: execute warm
 reset. R0B: execute hard reset

---
 drivers/hwmon/accton_t600_cpld.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/hwmon/accton_t600_cpld.c b/drivers/hwmon/accton_t600_cpld.c
index bac97d5..c636f63 100644
--- a/drivers/hwmon/accton_t600_cpld.c
+++ b/drivers/hwmon/accton_t600_cpld.c
@@ -301,6 +301,21 @@ static ssize_t reset(struct device *dev, struct device_attribute *da,
         return -EINVAL;
     }
 
+    /* check HW blade version.
+     * R0A: Not support Hard reset. We change to Warm reset to restart the system.
+     */
+    mutex_lock(&data->update_lock);
+    status = t600_cpld_read(client, BOARD_REV_REG);
+    if (unlikely(status < 0)) {
+        goto exit;
+    }
+    mutex_unlock(&data->update_lock);
+    if ((status & 0x7) == 0) /* R0A: 0x0, R0B: 0x1 */
+    {
+        if (reset == 1 )
+            reset = 2;   /* change HARD reset to WARM reset */
+    }
+
     mutex_lock(&data->update_lock);
     if (reset == 1) {
         /* RESET_ALL when SW hard reset */
-- 
1.9.1

